File: 프로그래머스/2/42587. 프로세스/프로세스.py
from collections import deque
import heapq
def solution(priorities, location):
    answer = 0
    queue = deque()
    maxPQ = []
    
    for i,p in enumerate(priorities):
        queue.append([i,p]) # index, priority
        heapq.heappush(maxPQ,-p)

    while maxPQ:
        targetP = -heapq.heappop(maxPQ)

        
        while queue[0][1] != targetP:
            elem = queue.popleft()
            queue.append(elem)

        answer += 1
        i,p = queue.popleft()
        if i == location:
            break
        

    return answer

File: 프로그래머스/2/42587. 프로세스/프로세스.java
import java.util.*;
class Process {
    int id;
    int priority;
    
    public Process(int id, int priority){
        this.id = id;
        this.priority = priority;
    }
}

class Solution {
    public int solution(int[] priorities, int location) {
        PriorityQueue<Integer> pq = new PriorityQueue<>((a,b) -> b - a);
        Deque<Process> q = new ArrayDeque<>();
        
        for(int i = 0; i < priorities.length; i++){
            pq.add(priorities[i]);
            q.addLast(new Process(i, priorities[i]));
        }
        
        int answer = 0;
        while(true){
            Process pr = q.removeFirst();
            if(pr.priority == pq.peek()){
                pq.poll();
                answer += 1;                
                if(pr.id == location) break;
            }
            else{
                q.addLast(pr);
            }
        }
        
        return answer;
    }
}